src/java/org/apache/cassandra/index/sai/plan/QueryController.java

```diff
@@ -206,103 +214,118 @@ public UnfilteredRowIterator getPartition(PrimaryKey key, ReadExecutionControlle
-                                                                                           RangeIterator<PrimaryKey> it = buildIterator(op, e.getValue(), isAnnHybridSearch);
-                                                                                           if (isAnnHybridSearch)
-                                                                                               it = reorderAndLimitBy(it, e.getKey(), annExpressionInHybridSearch);
-                                                                                           return it;
-                                                                                       })
-                                                                                       .collect(Collectors.toList());
-
-            Iterable<RangeIterator<PrimaryKey>> allIntersections = Iterables.concat(sstableIntersections, memtableIntersections);
-
-            queryContext.sstablesHit += queryView.referencedIndexes
-                                        .stream()
-                                        .map(SSTableIndex::getSSTable).collect(Collectors.toSet()).size();
-            queryContext.checkpoint();
-            RangeIterator<PrimaryKey> union = RangeUnionIterator.build(allIntersections);
-            return new CheckpointingIterator<>(union,
-                                               queryView.referencedIndexes,
-                                               annQueryViewInHybridSearch == null ? Collections.emptySet() : annQueryViewInHybridSearch.referencedIndexes,
-                                               queryContext);
+            return TermIterator.build(sstableIntersections, memtableResults, queryView.referencedIndexes, queryContext);
         }
         catch (Throwable t)
         {
             // all sstable indexes in view have been referenced, need to clean up when exception is thrown
             queryView.referencedIndexes.forEach(SSTableIndex::release);
-            // if ANN sstable indexes are referenced separately, release them
-            if (annQueryViewInHybridSearch != null)
-                annQueryViewInHybridSearch.referencedIndexes.forEach(SSTableIndex::release);
-
             throw t;
         }
     }
 
-    private RangeIterator<PrimaryKey> convertToPrimaryKeyIterator(PrimaryKeyMap.Factory pkFactory, RangeIterator<Long> sstableRowIdsIterator)
+    // This is a hybrid query. We apply all other predicates before ordering and limiting.
+    public RangeIterator getTopKRows(RangeIterator source, RowFilter.Expression expression)
     {
-        if (sstableRowIdsIterator.getCount() <= 0)
-            return RangeIterator.emptyKeys();
-
-        PrimaryKeyMap primaryKeyMap = pkFactory.newPerSSTablePrimaryKeyMap();
-        return SSTableRowIdKeyRangeIterator.create(primaryKeyMap, queryContext, sstableRowIdsIterator);
+        // TODO find way to test different chunk sizes. Found a fundamental flaw with 1, so it'd be good to continue
```

Comment:
fyi I've been using VSTODO to make it easier to see our TODOs vs the dozens that already exist in C*

Reply:
lol, I incorrectly assumed that was a VS Code version of TODO. I now see that VS == vector search. I'll update these.

Reply:
:)